core/services/ocr2/plugins/directrequestocr/reporting.go

```diff
@@ -0,0 +1,225 @@
+	if err != nil {
+		return nil, err
+	}
+
+	queryProto := Query{}
+	for _, result := range results {
+		queryProto.RequestIDs = append(queryProto.RequestIDs, result.ContractRequestID[:])
+	}
+	return proto.Marshal(&queryProto)
+}
+
+// Observation() complies with ReportingPlugin
+func (r *directRequestReporting) Observation(ctx context.Context, ts types.ReportTimestamp, query types.Query) (types.Observation, error) {
+	r.logger.Debug("directRequestReporting Observation phase", commontypes.LogFields{
+		"epoch": ts.Epoch,
+		"round": ts.Round,
+	})
+
+	queryProto := &Query{}
+	err := proto.Unmarshal(query, queryProto)
+	if err != nil {
+		return nil, err
+	}
+
+	observationProto := Observation{}
+	for _, id := range queryProto.RequestIDs {
+		localResult, _ := r.pluginORM.FindById(sliceToByte32(id))
+		if localResult.State == directrequestocr.RESULT_READY {
+			resultProto := ProcessedRequest{
+				RequestID: localResult.ContractRequestID[:],
+				Result:    localResult.Result,
+				Error:     []byte(localResult.Error),
+			}
+			observationProto.ProcessedRequests = append(observationProto.ProcessedRequests, &resultProto)
+		}
+	}
+	r.logger.Debug("directRequestReporting Observation phase done", commontypes.LogFields{
+		"nReadyRequests": len(observationProto.ProcessedRequests),
+	})
+
+	return proto.Marshal(&observationProto)
+}
+
+// Report() complies with ReportingPlugin
+func (r *directRequestReporting) Report(ctx context.Context, ts types.ReportTimestamp, query types.Query, obs []types.AttributedObservation) (bool, types.Report, error) {
+	r.logger.Debug("directRequestReporting Report phase", commontypes.LogFields{
+		"epoch":         ts.Epoch,
+		"round":         ts.Round,
+		"nObservations": len(obs),
+	})
+
+	queryProto := &Query{}
+	err := proto.Unmarshal(query, queryProto)
+	if err != nil {
+		r.logger.Error("directRequestReporting Report phase unable to decode query!",
+			commontypes.LogFields{"err": err})
+		return false, nil, err
+	}
+
+	reqIdToResultList := make(map[string][]*ProcessedRequest)
+	for _, id := range queryProto.RequestIDs {
+		reqIdToResultList[string(id)] = []*ProcessedRequest{}
+	}
+
+	for _, ob := range obs {
+		observationProto := &Observation{}
+		err = proto.Unmarshal(ob.Observation, observationProto)
+		if err != nil {
+			r.logger.Error("directRequestReporting Report phase unable to decode observation!",
+				commontypes.LogFields{"err": err, "observer": ob.Observer})
+			continue
+		}
+		for _, res := range observationProto.ProcessedRequests {
+			id := string(res.RequestID)
+			if val, ok := reqIdToResultList[id]; ok {
+				reqIdToResultList[id] = append(val, res)
+			}
+		}
+	}
+
+	// TODO make aggregation modular and configurable with Median as default.
+	// https://app.shortcut.com/chainlinklabs/story/56740/modular-aggregation
+	var aggregated []*ProcessedRequest
+	for _, obsArr := range reqIdToResultList {
+		if len(obsArr) >= 3 {
```

Comment:
We do exactly that - first we group all observations by request ID in the map reqIdToResultList. Then, for each request separately, we see if we have enough observations and only then include the aggregated results in the report. For now, the minimum is hardcoded to 3 and the aggregation is a trivial "take first observation". It will be implemented properly in my next PR, as indicated by the TODO comment above.

Reply:
I suspected that 3 is this threshold but it is confusing.
Let's at least replace it with some constant whose name denotes its role and leave a TODO to later get rid of the constant.

Reply:
I'll add a constant. The story mentioned in the existing TODO takes care of making everything configurable.